packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts

```diff
@@ -0,0 +1,218 @@
+import { FileNode, FileStatNode, FileTreeModel } from '@theia/filesystem/lib/browser';
+import {
+    ApplicationShell,
+    CompositeTreeNode,
+    open,
+    NavigatableWidget,
+    OpenerService,
+    SelectableTreeNode,
+    TreeNode,
+    Widget,
+    ExpandableTreeNode
+} from '@theia/core/lib/browser';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import debounce = require('@theia/core/shared/lodash.debounce');
+import { DisposableCollection } from '@theia/core/lib/common';
+
+export interface OpenEditorNode extends FileStatNode {
+    widget: Widget;
+};
+
+export namespace OpenEditorNode {
+    export function is(node: object | undefined): node is OpenEditorNode {
+        return FileStatNode.is(node) && 'widget' in node;
+    }
+}
+
+@injectable()
+export class OpenEditorsModel extends FileTreeModel {
+    static GROUP_NODE_ID_PREFIX = 'group-node';
+    static AREA_NODE_ID_PREFIX = 'area-node';
+
+    @inject(ApplicationShell) protected readonly applicationShell: ApplicationShell;
+    @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
+    @inject(OpenerService) protected readonly openerService: OpenerService;
+
+    protected toDisposeOnPreviewWidgetReplaced = new DisposableCollection();
+
+    protected _editorWidgetsByArea = new Map<ApplicationShell.Area, NavigatableWidget[]>();
+
+    get editorWidgets(): NavigatableWidget[] {
+        let editorWidgets: NavigatableWidget[] = [];
+        this._editorWidgetsByArea.forEach(widgets => {
+            editorWidgets = [...editorWidgets, ...widgets];
+        });
+        return editorWidgets;
+    }
+
+    getEditorWidgetsByGroup(id: ApplicationShell.Area | number): NavigatableWidget[] | undefined {
+        const idAsNum = Number(id);
+        if (!isNaN(idAsNum)) {
+            return this._editorWidgetsByGroup.get(idAsNum);
+        }
+        return this._editorWidgetsByArea.get(id as ApplicationShell.Area);
+    }
+
+    protected _editorWidgetsByGroup = new Map<number, NavigatableWidget[]>();
+
+    @postConstruct()
+    protected init(): void {
+        super.init();
+        this.setupHandlers();
+        this.initializeRoot();
+    }
+
+    protected setupHandlers(): void {
+        this.toDispose.push(this.selectionService.onSelectionChanged(selection => {
+            const { widget } = (selection[0] as OpenEditorNode);
+            this.applicationShell.activateWidget(widget.id);
+        }));
+        this.toDispose.push(this.applicationShell.onDidChangeCurrentWidget(async ({ newValue }) => {
+            const nodeToSelect = this.tree.getNode(newValue?.id) as SelectableTreeNode;
+            if (nodeToSelect) {
+                this.selectNode(nodeToSelect);
+            }
+        }));
+        this.toDispose.push(this.workspaceService.onWorkspaceChanged(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.workspaceService.onWorkspaceLocationChanged(() => this.updateOpenWidgets()));
```

Comment:
I see how these events change the decorations, but I'm not sure how they would affect widgets that are open?

Reply:
Good catch, I've removed these